### PR TITLE
gate: add metric for wait

### PR DESCRIPTION
**Pull request by paveldroo**
Add a metric that shows how long we need to wait before a request in the
remote read handler gets executed. It is useful to have it to understand
whether the concurrency limit needs to be lifted or not.

Closes https://github.com/prometheus/prometheus/issues/11365.

Signed-off-by: Pavel Druzhinin <paveldroo@gmail.com>

### CODE REVIEW

File: storage/remote/read_handler.go [hunk 1]
```diff
@@ -48,11 +50,17 @@ type readHandler struct {
...
-		remoteReadGate:            gate.New(remoteReadConcurrencyLimit),
+		logger:                logger,
+		queryable:             queryable,
+		config:                config,
+		remoteReadSampleLimit: remoteReadSampleLimit,
+		remoteReadGate: gate.New(remoteReadConcurrencyLimit, promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace:   "prometheus",
+			Subsystem:   "api",
+			Help:        "Total number of seconds spent waiting at the query gate before remote read query execution",
+			Name:        "prometheus_api_remote_read_wait_duration_seconds_total",
```
Comment (cstyan) [inline storage/remote/read_handler.go:61]: the earlier namespace and subsystem values will be prefixed so this name just needs to be `remote_read_wait_duration_seconds_total`

File: storage/remote/read_handler.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/prometheus/client_golang/prometheus/promauto"
```
Comment (cstyan) [inline storage/remote/read_handler.go:23]: this can move down to near the other `client_golang` import

File: util/gate/gate.go [hunk 3]
```diff
@@ -13,23 +13,36 @@
...
 type Gate struct {
-	ch chan struct{}
+	ch           chan struct{}
+	waitDuration prometheus.Counter
 }
 
 // New returns a query gate that limits the number of queries
 // being concurrently executed.
-func New(length int) *Gate {
+func New(length int, waitDuration prometheus.Counter) *Gate {
```
Comment (cstyan) [inline util/gate/gate.go:31]: I think the metric parameter here should be a pointer :thinking:

File: storage/remote/read_handler.go (hunk 1 above)
Comment (paveldroo) [reply storage/remote/read_handler.go]: Nice one.

File: storage/remote/read_handler.go (hunk 2 above)
Comment (paveldroo) [reply storage/remote/read_handler.go]: Done.

File: util/gate/gate.go (hunk 3 above)
Comment (paveldroo) [reply util/gate/gate.go]: I tried to rethink the suggestion once again, and it seems to have worked out. You can review it in this commit: https://github.com/prometheus/prometheus/pull/13565/commits/f223bfbaa411bd6a3c428e2e64e800593061cda2.

If the quantity and nature of changes are not entirely acceptable, everting the commit would be convenient.

PS Sorry if my previous comment is too silly :)

Comment (cstyan) [reply util/gate/gate.go]: Sorry, I made this comment just based on how we do similar "passing of common metrics", for example [here](https://github.com/prometheus/prometheus/blob/main/storage/remote/queue_manager.go#L443) where we pass a pointer to a struct of metrics. I still think the pointer is the right way to go :thinking:

Comment (paveldroo) [reply util/gate/gate.go]: Understood, anyway, it's implemented with a pointer now, as you suggested, and it looks not bad IMO. That was an interesting indication.
Then I'll be waiting for further review and approvals.
Thanks.

**Comment by cstyan** [comment]
The only other suggestion I would have would be to make the metric a package global, especially when using promauto. Alternatively we could replace promauto usage with regular metrics registration (and pass in a registerer). It should be safe as is since we only call `NewReadHandler` when initializing the API, which itself should only once.

I'd prefer to let someone more familiar with remote_read merge but they may all be on PTO for a bit, I'll check.

File: storage/remote/read_handler.go [hunk 4]
```diff
@@ -48,31 +47,31 @@ type readHandler struct {
...
 		remoteReadMaxBytesInFrame: remoteReadMaxBytesInFrame,
-		marshalPool:               &sync.Pool{},
-
 		queries: prometheus.NewGauge(prometheus.GaugeOpts{
 			Namespace: "prometheus",
 			Subsystem: "api", // TODO: changes to storage in Prometheus 3.0.
 			Name:      "remote_read_queries",
 			Help:      "The current number of remote read queries being executed or waiting.",
 		}),
+		marshalPool: &sync.Pool{},
```
Comment (paveldroo) [inline storage/remote/read_handler.go:71]: Changed the order of fields in accordance with the struct definition

**Comment by paveldroo** [comment]
> The only other suggestion I would have would be to make the metric a package global, especially when using promauto. Alternatively we could replace promauto usage with regular metrics registration (and pass in a registerer). It should be safe as is since we only call `NewReadHandler` when initializing the API, which itself should only once.
> 
> I'd prefer to let someone more familiar with remote_read merge but they may all be on PTO for a bit, I'll check.

I tried to implement your advice here https://github.com/prometheus/prometheus/pull/13565/commits/0ef4d8295666997ffc7b3eab0487207c25d3beb6, could you please check if I understood you correctly?

P.S. sorry for the force-pushes, there was a confusion with corporate email in commits.

File: storage/remote/read_handler.go [hunk 5]
```diff
@@ -47,24 +47,31 @@ type readHandler struct {
 // NewReadHandler creates a http.Handler that accepts remote read requests and
 // writes them to the provided queryable.
 func NewReadHandler(logger log.Logger, r prometheus.Registerer, queryable storage.SampleAndChunkQueryable, config func() config.Config, remoteReadSampleLimit, remoteReadConcurrencyLimit, remoteReadMaxBytesInFrame int) http.Handler {
+	waitDuration := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace:   "prometheus",
+		Subsystem:   "api",
+		Help:        "Total number of seconds spent waiting at the query gate before remote read query execution",
+		Name:        "remote_read_wait_duration_seconds_total",
+		ConstLabels: prometheus.Labels{"name": "read_handler"},
```
Comment (cstyan) [inline storage/remote/read_handler.go:55]: do we need this label? if the metric name is unique then we should be fine

Comment (paveldroo) [reply storage/remote/read_handler.go]: Sounds good, removed it.

**Comment by paveldroo** [comment]
kindly reminder

**Comment by bboreham** [comment]
[Reviewed at the bug-scrub]

The counter added here will let observers see the rate with which total waiting time is increasing.
This does not seem to address the request in #11356:

> if we are close to that limit then currently it is impossible to know that

Am I missing something?  How would you use this metric to achieve that?

**Comment by paveldroo** [comment]
> [Reviewed at the bug-scrub]
> 
> The counter added here will let observers see the rate with which total waiting time is increasing. This does not seem to address the request in #11356:
> 
> > if we are close to that limit then currently it is impossible to know that
> 
> Am I missing something? How would you use this metric to achieve that?

Fair enough, missed that point. What do you think about [concurrencyLimitHit metric](https://github.com/prometheus/prometheus/pull/13565/commits/07879e5417c70151248b4db84bc658d1fd260676)?
Or do you think we should be aware in advance that we are approaching the limit?

Thanks for your reply! And great talk about Looser Trees at GopherCon btw :)

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

@bboreham What do you think about the latest question from Pavel?

**Comment by bboreham** [comment]
I think the code has been substantially changed since I last looked at it; I don't think it matches the PR description any more.

I don't think I should be the person making the case that this solution is correct.  Please write down how someone would use this metric to address the problem it claims to be solving.

**Comment by paveldroo** [comment]
I agree that if no one else has shown interest in moving this improvement to production after such a long time, 
we should reconsider the necessity of this enhancement. Personally, I see this as purely a technical task for me and, unfortunately, I can't fully assess the business value of this feature.

I suggest either continuing work on this issue or closing it, but we shouldn't leave it hanging with the "help wanted" & "low hanging fruit" tags.

**Comment by bboreham** [comment]
I think we should close this then.

To be clear: I'm not saying this PR is bad, but as a reviewer I couldn't see how it would be used, and no explanation was provided.

**Pull request closed without merging**